### PR TITLE
Refactor the State class

### PR DESCRIPTION
**Pull request by xinyuhwang**

### CODE REVIEW

File: ChessStatePattern/src/states/State.java [hunk 1]
```diff
@@ -1,16 +1,36 @@
 package states;
 
 import main.*;
+import states.commons.Move;
+
+import java.util.ArrayList;
+import java.util.List;
 
 public abstract class State {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java:10]: Blank line, is not necessary.

File: ChessStatePattern/src/states/State.java [hunk 2]
```diff
@@ -1,16 +1,36 @@
 package states;
 
 import main.*;
+import states.commons.Move;
+
+import java.util.ArrayList;
+import java.util.List;
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java:8]: We need a Javadoc comment

File: ChessStatePattern/src/states/State.java [hunk 3]
```diff
@@ -1,16 +1,36 @@
...
 	ChessGame game;
-	
+	String activePlayer;
+	List<Move> moveHistory;
+
 	public State(ChessGame game) {
 		this.game = game;
+		moveHistory = new ArrayList<>();
+	}
+
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java:15]: Javadoc comment is needed

File: ChessStatePattern/src/states/State.java [hunk 4]
```diff
@@ -1,16 +1,36 @@
...
+	public void setActivePlayer(String playerName) {
+		this.activePlayer = playerName;
+	}
+
+	public String getActivePlayer() {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: Javadoc comment is needed

File: ChessStatePattern/src/states/State.java [hunk 5]
```diff
@@ -1,16 +1,36 @@
...
+		return this.activePlayer;
+	}
+
+	public void addMove(Move move) {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: Javadoc comment is needed

File: ChessStatePattern/src/states/State.java [hunk 6]
```diff
@@ -1,16 +1,36 @@
...
+		moveHistory.add(move);
+	}
+
+	public List<Move> getMoveHistory() {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: Javadoc comment is needed

File: ChessStatePattern/src/states/State.java [hunk 7]
```diff
@@ -1,16 +1,36 @@
...
+import states.commons.Move;
+
+import java.util.ArrayList;
+import java.util.List;
 
 public abstract class State {
 	
 	ChessGame game;
-	
+	String activePlayer;
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: Seems like you added `activePlayer` there and plan to track each player's actions,
but we need to ensure that each time switch the player action, this `activePlayer` needs to be updated. Will it be implemented in `setActivePlayer `?

File: ChessStatePattern/src/states/State.java (hunk 1 above)
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java:9]: Blank line, is not necessary.

File: ChessStatePattern/src/states/State.java [hunk 8]
```diff
@@ -1,16 +1,36 @@
 package states;
 
 import main.*;
+import states.commons.Move;
+
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: Blank line, is not necessary.

File: ChessStatePattern/src/states/State.java (hunk 2 above)
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: We need a Javadoc comment

File: ChessStatePattern/src/states/State.java [hunk 9]
```diff
@@ -1,16 +1,36 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
 public abstract class State {
 	
 	ChessGame game;
-	
+	String activePlayer;
+	List<Move> moveHistory;
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: In the state pattern, each state’s primary role is to define behavior, not to store or manage cross-state data. ChessGame as the main controller manages game-wide data, while state classes focus on defining state-specific behavior.

In this case, what do you think about moving `moveHistory` this to `ChessGame`?

File: ChessStatePattern/src/states/State.java [hunk 10]
```diff
@@ -1,16 +1,36 @@
...
+
+	public void addMove(Move move) {
+		moveHistory.add(move);
+	}
+
+	public List<Move> getMoveHistory() {
+		return moveHistory;
 	}
-	
-	//here is where you write the methods you will want overridden by each
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java:13]: Based on the professor provided hints, 
Which methods planned to be overridden by each specific state's behavior?

In my option, such as, `isGameOver`, `takeTurn`, etc?

File: ChessStatePattern/src/states/State.java [hunk 11]
```diff
@@ -1,16 +1,36 @@
 package states;
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java:1]: Left comments right here,

The PR title need to be modified that clear to appear what's the pr about, not only 'state'
For example, it's related to State class and Move class, the title should mention what's change/add/modify on them.

File: ChessStatePattern/src/states/commons/Move.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
+package states.commons;
+
```
Comment (ErdunE) [inline ChessStatePattern/src/states/commons/Move.java]: javadoc comment

File: ChessStatePattern/src/states/commons/Move.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+public class Move {
+    private int row;
+    private int col;
+    public Move(int row, int col) {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/commons/Move.java]: Just a discuss that
I was thinking if there need to set up edges condition and throw exception due to the board is 8 X 8
it's like once  0 > row or 7 < row or 0 > col or 7 < col

File: ChessStatePattern/src/states/commons/Move.java [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        this.row = row;
+        this.col = col;
+    }
+
+    public void setRow(int row) {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/commons/Move.java]: javadoc comment

File: ChessStatePattern/src/states/commons/Move.java [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+        this.row = row;
+    }
+
+    public void setCol(int col) {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/commons/Move.java]: javadoc comment

File: ChessStatePattern/src/states/commons/Move.java [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+        this.col = col;
+    }
+
+    public int getRow() {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/commons/Move.java]: javadoc comment

File: ChessStatePattern/src/states/commons/Move.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+        return row;
+    }
+
+    public int getCol() {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/commons/Move.java]: javadoc comment

**Review by ErdunE — COMMENTED** [review]
Summary：
1. Optimize the PR's title and description with more clear clarification of the changes.
2. Consider moving `moveHistory` to `ChessGame` for centralized management. In this way, each state can access and update a global move record instead of holding its own separate history.
3. Add edges conditions in the Move constructor to ensure each move represents a valid board position.
4. Adding comments on key methods to describe their purpose and usage will make the code clearer for other classmates

**Review by arete7-kai — APPROVED** [review]

**Review by arete7-kai — DISMISSED** [review]

File: ChessStatePattern/src/states/State.java [hunk 18]
```diff
@@ -1,16 +1,74 @@
...
+ * @author Xinyu Wang
+ * @version 10/31/2024
+ * The {@code State} class is the abstract superclass representing the state of the game.
+ */
 public abstract class State {
 	
 	ChessGame game;
-	
+	private boolean check;
+	private boolean checkMate;
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: `check` and `checkMate` the name make a bit confusing to use, due to they are only the states of boolean.
Suggest rename it as `isCheck` and `isCheckmate` for clarity

File: ChessStatePattern/src/states/State.java [hunk 19]
```diff
@@ -1,16 +1,74 @@
...
+	 */
+	public void setCheck() {
+		this.check = true;
+	}
+
+	/**
+	 * Set the state to normal play state.
+	 * Set the boolean value of check to false.
+	 */
+	public void setEscape() {
```
Comment (ErdunE) [inline ChessStatePattern/src/states/State.java]: The method name `setEscape()` is not quite clear enough to immediately see that its function is to cancel the  "Check" state. 
Consider changing it to `clearCheck()` or `setNormalPlay()` to express its function more clearly.

**Review by ErdunE — CHANGES_REQUESTED** [review]
Looks good to me overall, but I noticed a few minor changes that could improve clearer. Here’s a summary:

### Summary:
1. Rename `check` and `checkMate` to `isCheck` and `isCheckmate` for clarity, as they are boolean states.
2. Rename `setEscape()` to `clearCheck()` or `setNormalPlay()` for clearer indication of its function to cancel the “Check” state.
3. Optimize the PR title to better reflect its content.
4. Add a summary of class changes and optimizations in the PR description.

**Review by NEUNaveen — APPROVED** [review]

**Review by NEUNaveen — DISMISSED** [review]

**Review by ErdunE — APPROVED** [review]
Worked with Kai and Hailee in Zoom meeting together, and we all approved this after discussed

**Pull request merged**